segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+                                        startOffset.set(segment.getInfo().getStartOffset());
+                                        readLength.set((int) Math.min(segment.getInfo().getLength() - startOffset.get(), MAX_TOTAL_EVENT_SIZE));
+                                        return segment.read(startOffset.get(), readLength.get(), containerOperationTimeout);
+                                    }, executor)
+                                    .thenComposeAsync(rr -> AsyncReadResultProcessor.processAll(rr, executor, containerOperationTimeout), executor)
+                                    .thenApply(inputData -> deserializeEvents(inputData, startOffset.get(), readLength.get()));
+        }
+
+        private EventsReadAndTruncationPoints deserializeEvents(BufferView inputData, long startOffset, int maxLength) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long nextOffset = startOffset;
+            final long maxOffset = startOffset + maxLength;
+            BufferView.Reader input = inputData.getBufferViewReader();
+            try {
+                while (nextOffset < maxOffset && events.size() < getConfig().getMaxItemsAtOnce()) {
+                    ProcessorEventData event = ProcessorEventSerializer.deserializeEvent(input);
+                    events.add(event);
+                    // Update the offset to the beginning of the next event.
+                    nextOffset += HEADER_LENGTH + event.getLength();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            }
+
+            if (events.isEmpty()) {
+                throw new NoDataAvailableException();
+            }
+            return new EventsReadAndTruncationPoints(events);
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> applyProcessorHandler(EventsReadAndTruncationPoints readResult) {
+            return getHandler().apply(readResult.getProcessorEventsData()).thenApply(v -> readResult);
+        }
+
+        private CompletableFuture<Void> truncateInternalSegment(EventsReadAndTruncationPoints readResult, Timer iterationTime) {
+            long truncationOffset = this.segment.getInfo().getStartOffset() + readResult.getReadBytes();
+            return this.segment.truncate(truncationOffset, containerOperationTimeout)
+                    .thenAccept(v -> {
+                        // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                        this.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                        this.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                    });
+        }
+    }
+
+    @Data
+    private static class EventsReadAndTruncationPoints {
```

Comment:
Can this be moved into the eventProcessor itself, so that all data about particular event processor is in one place on one object.

where is EventsReadAndTruncationPoints used? Looks like I missed that part - so some comments will be helpful.

Reply:
Done. It is just a convenience class to encapsulate read events and truncation info.